embedded-graphics/src/primitives/rectangle.rs

```diff
@@ -138,6 +138,78 @@ impl Rectangle {
             None
         }
     }
+
+    /// Returns a new `Rectangle` containing the intersection of `self` and `other`.
+    ///
+    /// If no intersection is present, this method will return `None`.
+    ///
+    /// # Examples
+    ///
+    /// This example draws two rectangles to a mock display using the `.` character, along with
+    /// their intersection shown with `#` characters.
+    ///
+    /// ```rust
+    /// use embedded_graphics::{
+    ///     mock_display::MockDisplay, pixelcolor::BinaryColor, prelude::*, primitives::Rectangle,
+    ///     style::PrimitiveStyle,
+    /// };
+    ///
+    /// let mut display = MockDisplay::new();
+    ///
+    /// let rect1 = Rectangle::new(Point::zero(), Size::new(7, 8));
+    /// let rect2 = Rectangle::new(Point::new(2, 3), Size::new(10, 7));
+    ///
+    /// // If there is no intersection, this unwrap() will panic
+    /// let intersection = rect1.intersection(&rect2).unwrap();
+    ///
+    /// rect1
+    ///     .into_styled(PrimitiveStyle::with_stroke(BinaryColor::Off, 1))
+    ///     .draw(&mut display)?;
+    ///
+    /// rect2
+    ///     .into_styled(PrimitiveStyle::with_stroke(BinaryColor::Off, 1))
+    ///     .draw(&mut display)?;
+    ///
+    /// intersection
+    ///     .into_styled(PrimitiveStyle::with_stroke(BinaryColor::On, 1))
+    ///     .draw(&mut display)?;
+    ///
+    /// assert_eq!(
+    ///     display,
+    ///     MockDisplay::from_pattern(&[
+    ///         ".......     ",
+    ///         ".     .     ",
+    ///         ".     .     ",
+    ///         ". #####.....",
+    ///         ". #   #    .",
+    ///         ". #   #    .",
+    ///         ". #   #    .",
+    ///         "..#####    .",
+    ///         "  .        .",
+    ///         "  ..........",
+    ///     ])
+    /// );
+    /// # Ok::<(), core::convert::Infallible>(())
+    /// ```
+    pub fn intersection(&self, other: &Rectangle) -> Option<Rectangle> {
```

Comment:
Should we change this to simply return a `Rectangle`? Because a `Rectangle`can already be zero sized adding an `Option` should be unnecessary. The `fill_solid` example demonstrates this problem IMO, because it needs to use two if statements.

Reply:
It could indeed, with checking for an intersection changing to `intersection.size == Size::zero()` which is pretty clean. In the case of no intersection at all, what should the `top_left` component of the zero-sized rectangle be? `Point::zero()`?

Reply:
>  In the case of no intersection at all, what should the `top_left` component of the zero-sized rectangle be? `Point::zero()`?

Yes, I think that's the best option.

Reply:
Changes pushed :ok_hand: